inferece: remove `CachedMethodTable`

aviatesk opened the pull request:
Previously the method lookup result was created per frame and so the
look cache hasn't been use that much. With this change the cache is
created per inference, and so the cached result will be used when we
already saw the same match in the same inference shot, and it may speed
up the lookup time a bit.

This commit also setups new `AbstractInterpreter` interface `get_method_lookup_cache`
which specifies what method lookup cache is used by each `AbstractInterpreter`.
`NativeInterpreter` creates a cache per inference, and so it is valid
since lookup is done in the same world age in the same inference shot.
External `AbstractInterpreter` doesn't opt into this cache by default,
and its behavior won't change in anyway.

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

vtjnash reviewed (commented):
Is this that helpful anymore? For concrete types, we already cache those in a hash table at the next level

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/44240#issuecomment-1044609789) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/e62b009_vs_daa0849/report.md).

aviatesk commented:
Lol, the benchmark shows that this change make it less efficient both in time and space...
@nanosoldier `runbenchmarks("inference", vs=":master")`

vtjnash commented:
yeah, we might want to do this and have only non-concrete types put here, otherwise we may end up doing the lookup twice in the cache-miss case.

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

aviatesk commented:
ok, I really couldn't confirm any performance improvement in `CachedMethodTable` in the current infrastructure.
Now I'd like to propose to eliminate that entirely, and it should save a bit of space:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/44240#issuecomment-1046498054) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/dd61bc4_vs_f879d20/report.md).

vtjnash commented on base/compiler/types.jl:
```diff
@@ -111,6 +111,23 @@ decode_effects_override(e::UInt8) =
+    return (match::MethodMatch, state)
+end
+getindex(result::MethodLookupResult, idx::Int) = getindex(result.matches, idx)::MethodMatch
+const MethodLookupCache = IdDict{Any, Union{Missing, MethodLookupResult}}
```
seems unused?
```suggestion
```

aviatesk commented on base/compiler/types.jl:
```diff
@@ -263,9 +280,8 @@ struct NativeInterpreter <: AbstractInterpreter
-
         return new(
-            # Initially empty cache
+            # Initially empty caches
```
```suggestion
            # Initially empty cache
```

aviatesk committed:
inference: remove `CachedMethodTable`

Since we couldn't confirm any performance benefit from `CachedMethodTable`
in the current infrastructure (see the benchmark results in #44240), now
I'd like to propose to eliminate that entirely and save a bit of space.
